framework/src/auxkernels/MortarNodalAuxKernel.C

```diff
@@ -0,0 +1,75 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MortarNodalAuxKernel.h"
+#include "MooseVariableField.h"
+#include "MortarUtils.h"
+
+#include "libmesh/quadrature.h"
+
+template <typename ComputeValueType>
+InputParameters
+MortarNodalAuxKernelTempl<ComputeValueType>::validParams()
+{
+  InputParameters params = AuxKernelTempl<ComputeValueType>::validParams();
+  params += MortarInterface::validParams();
+  params.set<bool>("use_displaced_mesh") = true;
+  return params;
+}
+
+template <typename ComputeValueType>
+MortarNodalAuxKernelTempl<ComputeValueType>::MortarNodalAuxKernelTempl(
+    const InputParameters & parameters)
+  : AuxKernelTempl<ComputeValueType>(parameters),
+    MortarInterface(this),
+    _displaced(this->template getParam<bool>("use_displaced_mesh")),
+    _fe_problem(*this->template getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
+    _msm_volume(0)
+{
+  if (!isNodal())
+    paramError("variable", "MortarNodalAuxKernel derivatives populate nodal aux variables only.");
+}
+
+template <typename ComputeValueType>
+void
+MortarNodalAuxKernelTempl<ComputeValueType>::compute()
+{
+  ComputeValueType value(0);
+  Real total_volume = 0;
+
+  const auto & its = amg().secondariesToMortarSegments(*_current_node);
+  std::array<MortarNodalAuxKernelTempl<ComputeValueType> *, 1> consumers = {this};
+
+  auto act_functor = [&value, &total_volume, this]()
+  {
+    _msm_volume = 0;
+    value += computeValue();
+    total_volume += _msm_volume;
+  };
+
+  Moose::Mortar::loopOverMortarSegments(
+      its, _assembly, _subproblem, _fe_problem, amg(), _displaced, consumers, act_functor);
```

Comment:
This method is computing the nodal value (using the `act_functor` among others): Do you think its name should be changed to reflect this? Say, something like `computeValueLoopingOverMortarSegments`. Maybe the answer is just that name is too long.

Reply:
This particular call to `loopOverMortarSegments` will `computeValue` for an aux kernel. But in `ComputeMortarFunctor` we call `loopOverMortarSegments` with an action functor that calls `computeResidual` or `computeJacobian`. The flexibility in `loopOverMortarSegments` is that we do not know what the user of the function will be doing in their action functor

Reply:
Makes sense.